content/blog/2024/2024-02-13-apache-http-client-config.md

```diff
@@ -0,0 +1,267 @@
+
+Typically, a custom response interceptor implements the `HttpResponseInterceptor` interface and overrides the `process()` method. Inside this method, developers can access and manipulate the HTTP response object, such as modifying headers, inspecting status codes, or extracting content.
+
+Custom response interceptors are useful for tasks like logging responses, handling error conditions, extracting specific information from responses, or performing additional processing before passing the response back to the client code. They provide flexibility and extensibility to tailor the behavior of HTTP responses according to application needs.
+
+Let us implement a response interceptor:
+```java
+@Slf4j
+public class CustomHttpResponseInterceptor implements HttpResponseInterceptor {
+  @Override
+  public void process(HttpResponse response, EntityDetails entity, HttpContext context)
+      throws HttpException, IOException {
+    log.debug("Got {} response from server.", response.getCode());
+  }
+}
+```
+
+The `CustomHttpResponseInterceptor` implements the `HttpResponseInterceptor` interface provided by Apache HttpClient. This interceptor is designed to intercept incoming HTTP responses before they are returned to the client. In the `process()` method the interceptor logs the status code of the response object.
+
+Now let us build a HTTP client using this request interceptor:
+``` java
+HttpResponseInterceptor interceptor = new CustomHttpResponseInterceptor();
+HttpClientBuilder builder = HttpClients.custom();
+CloseableHttpClient client = builder.addResponseInterceptorFirst(interceptor).build();
+
+```
+In this code snippet, we first create an instance of `CustomHttpResponseInterceptor` to handle incoming HTTP responses. Then we build the client using `HttpClientBuilder`.
+
+Then we call `addResponseInterceptorFirst()` method passing the interceptor object as an argument. This method adds the `interceptor` as the first response interceptor in the chain of interceptors. It also has a method `addResponseInterceptorLast()` to add the interceptor at the end.
+
+By adding the interceptor first, it ensures that the response code logged before the response is manipulated further. Finally, it calls the `build()` method to create the `CloseableHttpClient` instance with the configured response interceptor.
+
+## Configuring Execution Interceptors
+On the other hand, an execution interceptor allows developers to intercept the execution of HTTP requests and responses. It operates at a lower level compared to request interceptors and can intercept various stages of the request execution process, such as before sending the request, after receiving the response, or when an exception occurs during execution. Execution interceptors can be used for tasks like logging, caching, or error handling. They provide finer-grained control over the HTTP request execution process and enable developers to implement custom logic based on specific requirements or conditions.
```

Comment:
This sentence doesn't clearly point out the difference to the Request Interceptor which as we wrote before also intercepts the request *before* it's sent. Maybe it's an idea to add a sequence diagram to this section?

Reply:
Simplified the text now.